### PR TITLE
chat: dynamic number based behavior for CLI; also put unread convs first

### PR DESCRIPTION
**Pull request by songgao**
This PR replaces `--number` with something more dynamic. Also it changes so that unread conversations are listed first. This also allows to have better default behavior of naked `keybase chat list`. It tries to list 2 additional read conversations after unread conversations as a subtle confirmation for user to know if there are any additional unread conversations.

See https://keybase.atlassian.net/browse/CORE-3827 for more details.

Here are some sample outputs:

```
songgao at Song-ThinkPad in ~GOPATH/src/github.com/keybase/client/go/keybase (songgao/CORE-3827●)
$ ./keybase chat list 
▶ WARNING Running in devel mode
[1] * t_alice,t_bob,test2           [test2 5d] hello!!!                                                                        
[2] * t_bob,t_charlie,test2         [test2 5d] test                                                                            
[3] * t_bob,t_charlie,t_doug,test2  [test2 5d] test                                                                            
[4] * t_doug,test2                  [test2 6d] doug                                                                            
[5]   t_alice,t_bob,t_charlie,test2 [test2 9h] hi                                                                              
[6]   test2                         [test2 9h] hi                                                                              

songgao at Song-ThinkPad in ~GOPATH/src/github.com/keybase/client/go/keybase (songgao/CORE-3827)
$ ./keybase chat list --at-most 2
▶ WARNING Running in devel mode
[1] * t_alice,t_bob,test2   [test2 5d] hello!!!                                                                                
[2] * t_bob,t_charlie,test2 [test2 5d] test                                                                                    

Number of conversations is capped by --at-most, so there might be more unread ones. Specify --at-most to a large number to fetch more.

songgao at Song-ThinkPad in ~GOPATH/src/github.com/keybase/client/go/keybase (songgao/CORE-3827)
$ ./keybase chat list --at-least 100
▶ WARNING Running in devel mode
[1] * t_alice,t_bob,test2           [test2 5d] hello!!!                                                                        
[2] * t_bob,t_charlie,test2         [test2 5d] test                                                                            
[3] * t_bob,t_charlie,t_doug,test2  [test2 5d] test                                                                            
[4] * t_doug,test2                  [test2 6d] doug                                                                            
[5]   t_alice,t_bob,t_charlie,test2 [test2 9h] hi                                                                              
[6]   test2                         [test2 9h] hi                                                                              
[7]   t_alice,test2                 [test2 3d] O_O                                                                             
[8]   t_charlie,test2               [test2 3d] >.<                                                                             
[9]   t_bob,test2                   [test2 3d] O_O                                                                             

```

r? @patrickxb @mmaxim @maxtaco Thanks!

### CODE REVIEW

**Comment by mmaxim** [comment]
As I mentioned in Slack, I think it is the wrong play to have unread on top, that doesn't appear to be how other chat apps work.

**Comment by songgao** [comment]
@mmaxim I ended up making a separate sub-command `keybase chat list-unread` for that dynamic number behavior, since the flags became complicated and sort of confusing. Let me know what you think!

File: go/client/cmd_chat_list.go [hunk 1]
```diff
@@ -20,13 +20,13 @@ type cmdChatList struct {
 func newCmdChatList(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
 	return cli.Command{
 		Name:         "list",
-		Usage:        "Show new messages in inbox.",
+		Usage:        "Show conversations in inbox, sorted by activities.",
```
Comment (mmaxim) [inline go/client/cmd_chat_list.go]: "sorted by activity"

File: protocol/avdl/chat1/local.avdl [hunk 2]
```diff
@@ -111,13 +111,27 @@ protocol local {
...
 
+  // This causes fetching to return N items, where N = IdeallyGetUnreadPlus +
+  // Unread, if AtLeast <= N <= AtMost, or one of the bounds if there are too
+  // many / too few unread items. i.e. (derived from chris's comment)
+  // collar(AtLeast, (IdeallyGetUnreadPlus + Unread), AtMost)
+  //
+  // By definition, one could use a same non-zero number for both AtLeast and
+  // AtMost to precisely control the number of items returned.
+  record NumLimit {
+    int IdeallyGetUnreadPlus;
```
Comment (mmaxim) [inline protocol/avdl/chat1/local.avdl]: Is there a better name we could use here? I have no idea what this means by just reading the name.

**Comment by mmaxim** [comment]
I really think we should get some sort of testing in for this too, some of this code is pretty complicated and would probably benefit from some test coverage.

**Comment by songgao** [comment]
@mmaxim ah OK! I'll add tests in service using the server mock in this PR then. Thanks!

File: protocol/avdl/chat1/local.avdl (hunk 2 above)
Comment (songgao) [inline protocol/avdl/chat1/local.avdl]: How about this?

```
record UnreadFirstNumLimit {
  int numRead;

  int atLeast;
  int atMost;
}

Comment (mmaxim) [inline protocol/avdl/chat1/local.avdl]: Sounds good!

**Comment by songgao** [comment]
I realized current chat mock for service tests only support one FakeUser. I needed to make it support multiple ones, so it's taking much longer than expected. But I'll try to get this done by sprint planning!

**Comment by songgao** [comment]
@mmaxim I added multi-user support for the mock, and added some tests for the dynamic number limits. PTAL thanks!

**Comment by songgao** [comment]
@maxtaco Could you give this another look before I merge it? Thanks!

**Pull request merged**